New "extend" keyword to colors.BoundaryNorm

fmaussion opened the pull request:
This is a follow-up to https://github.com/matplotlib/matplotlib/issues/4850

**Rationale**: when using BoundaryNorm with a continuous colormap, you would want the colors of the extensions to be distinct from the colors of the nearest box (see example below).

**Compatibility**: there is no backward compatibility issue since the default behavior is unchanged

**Limitations of the current implementation**: it adds a bit of an overhead of code (not too much but still). An alternative implementation would be to make a new class, e.g. ExtendedBoundaryNorm in order to take over this functionality. Another issue is that it adds a behavior to one of the implementations of the Norm interface but not to the others. I am not sure which of the other classes might benefit from such a keyword. 

**Example**

``` python
"""
Illustrate the use of BoundaryNorm wht the "extend" keyword
"""

from matplotlib import pyplot
import matplotlib as mpl

# Make a figure and axes with dimensions as desired.
fig = pyplot.figure(figsize=(8, 3))
ax1 = fig.add_axes([0.05, 0.80, 0.9, 0.15])
ax2 = fig.add_axes([0.05, 0.475, 0.9, 0.15])
ax3 = fig.add_axes([0.05, 0.15, 0.9, 0.15])

# Set the colormap and bounds
bounds = [-1, 2, 5, 7, 12, 15]
cmap = mpl.cm.get_cmap('viridis')

# Default behavior
norm = mpl.colors.BoundaryNorm(bounds, cmap.N)
cb1 = mpl.colorbar.ColorbarBase(ax1, cmap=cmap,
                                     norm=norm,
                                     extend='both',
                                     orientation='horizontal')
cb1.set_label('Default BoundaryNorm ouput');

# New behavior
norm = mpl.colors.BoundaryNorm(bounds, cmap.N, extend='both')
cb2 = mpl.colorbar.ColorbarBase(ax2, cmap=cmap,
                                     norm=norm,
                                     # I am still unhappy with this keyword
                                     # which is repetitive, but I see no way
                                     # around right now
                                     extend='both',
                                     orientation='horizontal')
cb2.set_label("With new extend='both' keyword");

# The rest remains unchanged
norm = mpl.colors.BoundaryNorm(bounds, cmap.N, extend='max')
cb3 = mpl.colorbar.ColorbarBase(ax3, cmap=cmap,
                                     norm=norm,
                                     extend='max',
                                     spacing='proportional',
                                     orientation='horizontal')
cb3.set_label("extend='max' and proportional spacing");

pyplot.show()
```

![figure_1](https://cloud.githubusercontent.com/assets/10050469/9713550/6991a71a-5552-11e5-994b-4c360692b97d.png)

tacaswell commented:
How is this different that the over/under functionality we already have?

fmaussion commented:
Do you mean Colormap.set_over() ? It is different in how BoundaryNorm() is choosing the colors *within* the levels' range. If you set_over(), the color of the last box before the extension will still be the last color of the colormap.

It is probable that I missed the way to do what I need without this PR, but I couldn't find any example in the gallery. @efiring seemed to agree on the rationale of this PR, but if proven useless I have no problem top close this PR (I just found it quite useful myself).

tacaswell commented:
Sorry, left a comment without reading enough context :disappointed: 

The point of this is to reserve levels _in the normalization_ to use as the over/under colors ?  I see how to do this using existing tools and it is annoying.

How does this interact with the clip kwarg?

fmaussion commented:
That's a good point: using clip=True and anything else than extend='neither' makes no sense, *unless* you want to clip the data and keep the normalized colors of the example 2 above, which is a very strange use case.

Currently something silently happens if you set extend and clip together, but it seems wrong. Should I force extend to "neither" if clip==True, or should I take care of the use-case above? 

(I hope its clear enough, otherwise I can provide an example)

tacaswell commented:
I am generally in favor of noisy exceptions (ex if `extend` is not `'neither'`, raise if `clip` is `True`).

Instead of calling this 'extend' in the normalization class, how about 'open_ends' or something like that?  It seems what this is really specifying is if there should be an implicit +/- inf on the edges of the boundary list.

fmaussion commented:
I agree with raising an exception. 

The reason for calling it `extend=` is that it really merges the concepts of `BoundaryNorm` and `Colorbar`. If you recall the gallery example http://matplotlib.org/examples/api/colorbar_only.html (example 3), the keywords "boundaries" and "ticks" are actually redundant (because `ColorbarBase` knows about the norm properties). 

Ideally, in my example above, the `extend` keyword in the call to `ColorbarBase` should be unnecessary. I *think* that it is possible to do this automatically but I didn't want to invest too much time in it without asking you first.

tacaswell commented:
That is exactly why I _don't_ want to call it extend.  My knee-jerk reaction to reduce the coupling between the normalization classes and the colorbar base class.  

I am in general very very wary of doing things automatically, but can be convinced otherwise.

fmaussion commented:
:D 

From my point-of-view (quite new to python after a long time with IDL), the example of http://matplotlib.org/examples/api/colorbar_only.html is horribly wordy. In fact, I assume that it was written in times before that the ColorbarBase base class was able to get all these infos from the normalization classes. It seems that the coupling you are complaining about already occurred ;).

For my own purposes I made a small library which adds this functionality and adds some higher level wrappers in order to prevent any mismatch between data, plot, and colorbar, which seems to be quite easy to occur with matplotlib. 

But one argument in favour of *decoupling* however is that various functions seem to react differently to the norm class. 

I'll misuse another example (http://matplotlib.org/examples/images_contours_and_fields/pcolormesh_levels.html). I just changed the levels in order to make use of extend='both':

```python
import matplotlib.pyplot as plt
from matplotlib.colors import BoundaryNorm
import numpy as np


# Make the data
dx, dy = 0.05, 0.05
y, x = np.mgrid[slice(1, 5 + dy, dy),
                slice(1, 5 + dx, dx)]
z = np.sin(x) ** 10 + np.cos(10 + y * x) * np.cos(x)
z = z[:-1, :-1]

# Z roughly varies between -1 and +1
# my levels are chosen so that the color bar should be extended
levels = [-0.8, -0.5, -0.2, 0.2, 0.5, 0.8]
cmap = plt.get_cmap('PiYG')
norm = BoundaryNorm(levels, ncolors=cmap.N, extend='both')

# Plot 1
plt.subplot(2, 1, 1)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar(extend='both')
plt.axis([x.min(), x.max(), y.min(), y.max()])
plt.title('pcolormesh with levels')

# Plot 2
plt.subplot(2, 1, 2)
plt.contourf(x[:-1, :-1] + dx / 2.,
             y[:-1, :-1] + dy / 2., z, levels=levels,
             cmap=cmap, extend='both')
plt.colorbar()
plt.title('contourf with levels')

plt.show()
```
![figure_1-1](https://cloud.githubusercontent.com/assets/10050469/9721236/828daeb8-5597-11e5-839d-b0ce72a83a7f.png)

The two plots are equivalent and *correct* but for contourf, the `extend` call occurs in `contourf`, while for `pcolormesh` it has to occur at the `colorbar` call.

If I make no call to `extend` at all but choose `clip=True` (previous):

```python
norm = BoundaryNorm(levels, ncolors=cmap.N, clip=True)

# Plot 1
plt.subplot(2, 1, 1)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar()
plt.axis([x.min(), x.max(), y.min(), y.max()])
plt.title('pcolormesh with levels')

# Plot 2
plt.subplot(2, 1, 2)
plt.contourf(x[:-1, :-1] + dx / 2.,
             y[:-1, :-1] + dy / 2., z, levels=levels,
             cmap=cmap)
plt.colorbar()
plt.title('contourf with levels')
```
![figure_1-2](https://cloud.githubusercontent.com/assets/10050469/9721392/0069362a-559a-11e5-8c4a-b333a89234e9.png)

It is amusing to see that contourf chooses different colors (somehow in accordance with my use-case for `extend`?) and marks out of bounds with white while pcolormesh is misleading since values above 0.8 get the same color as below.

**To conclude**: I'd prefer to have more control of the Normalize class on the plotting functions because the two are inherently related, BUT I can understand that this is going to be a mess if we start to change things...

fmaussion commented:
Sorry for my confusing post above. I'll get back to you tomorrow after thinking about all this a bit more ;-)

tacaswell commented:
In your second example, `pcolor` is doing what I expect for `clip=True` .  In the `contourf` case the white areas at the high/low ends I suspect are `np.nan` which is mapped to 'bad' which iirc maps to transparent.  The color that `contourf` hands to the normalization class is the center of the bands, not the edges, hence why it picks the same colors for the valid data in both cases.

In _this_ case where you are using boundary norms the tight coupling is natural, but say you want to use `contourf` with a non-linear color map, but with linear levels.

fmaussion commented:
Yes sorry, the contourf example was OT because it has nothing to do with the normalization. 

I see two solutions:

1. If we call the keyword "extend", it has to be taken into account by all functions that take `norm=*` as argument. Otherwise it is misleading, since all other properties (such as bounds, colors and clipping) are already derived from the normalization class as shown by the pcolormesh example above. I largely prefer this solution because it is consistent and less buggy. However, I have too little experience with mpl to estimate the amount of changes that it would represent. 

2. We call it something else and leave it as is it now. It would remain a minimal change and would maybe require a new gallery example to illustrate its use.

What do you think?

tacaswell commented:
I don't quite understand 1.  In both cases the plotting function takes in some data, does some computation, and then gets a scalar out.  This scaler is then passed to the `norm` to be re-scaled to `[0, 1]` (or `[0, N]`) and the rescaled value is then passed to `cmap` to be turned into an `RGBA` value.  I don't think there is any case where the plotting functions are introspecting anything about `norm`.  In the `pcolormesh` case each square is just going through that mapping process.

fmaussion commented:
Yes, ok. I don't understant all the details of what is going under the hood when a call to pyplot.colorbar is made:

```python
norm = BoundaryNorm(levels, ncolors=cmap.N, extend='both')

# Plot 1
plt.subplot(2, 1, 1)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar()
plt.title('pcolormesh, forgot extend to colorbar')

# Plot 2
plt.subplot(2, 1, 2)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar(extend='both')
plt.title('pcolormesh with levels')
```

![figure_1-5](https://cloud.githubusercontent.com/assets/10050469/9737820/bc487fbe-5648-11e5-9ec5-e358a42f68a4.png)

So the plot is fine with my new implementation, so it's really just about the colorbar. I guess that since the colorbar knows the levels I've chosen, it must know about the normalize object I gave to pcolormesh.

tacaswell commented:
You will have to ask @efiring about the colorbar code.

fmaussion commented:
Ok, the change needed to make it do what I needed was minimal, but I have too little overview of the whole thing to judge if it's a good thing or not. If find it more consistent since no double call is needed, but I can understand that it is quite a specific requirement of mine. 

By adding:
```python
if hasattr(norm, 'extend') and norm.extend != 'neither':
    extend = norm.extend
```
in ColorbarBase's __init__ I am forcing the keyword to what I want. If you agree on the "intrusion" I will add a few tests, if not we should decide on a name for the keyword.

My previous examples now look like this:

```python
""" Illustrate the use of BoundaryNorm wht the "extend" keyword """

import matplotlib.pyplot as plt
import matplotlib as mpl

# Make a figure and axes with dimensions as desired.
fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(6, 2))

# Set the colormap and bounds
bounds = [-1, 2, 5, 7, 12, 15]
cmap = mpl.cm.get_cmap('viridis')

# Default behavior
norm = mpl.colors.BoundaryNorm(bounds, cmap.N)
cb1 = mpl.colorbar.ColorbarBase(ax1, cmap=cmap, norm=norm, extend='both',
                                orientation='horizontal')
cb1.set_label('Default BoundaryNorm ouput');

# New behavior
norm = mpl.colors.BoundaryNorm(bounds, cmap.N, extend='both')
cb2 = mpl.colorbar.ColorbarBase(ax2, cmap=cmap, norm=norm,
                                orientation='horizontal')
cb2.set_label("With new extend='both' keyword");
plt.tight_layout()
plt.show()
```
![figure_1-6](https://cloud.githubusercontent.com/assets/10050469/9759633/b826b8be-56ee-11e5-8922-aca46c37519d.png)

```python
import matplotlib.pyplot as plt
from matplotlib.colors import BoundaryNorm
import numpy as np

# Make the data
dx, dy = 0.05, 0.05
y, x = np.mgrid[slice(1, 5 + dy, dy),
                slice(1, 5 + dx, dx)]
z = np.sin(x) ** 10 + np.cos(10 + y * x) * np.cos(x)
z = z[:-1, :-1]

# Z roughly varies between -1 and +1
# my levels are chosen so that the color bar should be extended
levels = [-0.8, -0.5, -0.2, 0.2, 0.5, 0.8]
cmap = plt.get_cmap('PiYG')
norm = BoundaryNorm(levels, ncolors=cmap.N, extend='both')

# Plot 1
plt.subplot(2, 1, 1)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar()
plt.title('setting extend=both is obsolete')

# Plot 2
plt.subplot(2, 1, 2)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar(extend='neither')
plt.title('extend=neither however is ignored')

plt.tight_layout()
plt.show()
```
![figure_1-7](https://cloud.githubusercontent.com/assets/10050469/9759635/c6929198-56ee-11e5-9aa9-3144a04ce2a1.png)

efiring commented:
Sorry to have let this slide for so long; I will try to spend some time on it today.  Something along these lines will be good, but I think there might be a better approach: a norm kwarg (or pair of kwargs) that changes the target range from 0-1 to some other range.  I think this might handle the use case for this PR in a very general way (e.g., by setting the target range to [0.1, 0.9]), and as a bonus, handle some other use cases at the same time.

tacaswell commented:
My drive-by comment on this is that there should be a way to do this without breaking any existing code.  If the user says `extend='neither'` then they should get neither.  I think this can be done by changing the default to `None`, if it is `None` then consult the `norm` then fall back to the current default.  If the user passes in a non-None value, respect it.

efiring commented:
What I'm suggesting does not inherently involve an "extend" kwarg, or any magic; it would be entirely explicit, and would have no effect on existing code.  It would be independent of the "extend" kwargs in colorbar and contourf.

tacaswell commented:
@efiring Any update on this?

efiring commented:
I'll make a point of getting back to this and other color questions no later than this weekend.  Thanks for the reminder.

fmaussion commented:
Sorry I was not able to be more helpful on this one. Let me know if I can do anything.

As a side note, xray already implements the logic I'm asking for:

```python
import xray
a = xray.DataArray([[1,2,3],[4,5,6]])
a.plot(levels=[2,3,4,5]);
```
Will produce (note the colorbar colors):
![xray_logic](https://cloud.githubusercontent.com/assets/10050469/10968497/dc1a5604-83c1-11e5-8e27-8c0000755f3c.png)

Maybe @shoyer can comment on this (no need to read all my lengthy examples, the original post will suffice).

efiring commented:
This looks consistent with the way `contourf` works.  I like it.  @fmaussion, would you make the one change suggested by @tacaswell, please? If I understand correctly, the point is that if the call to `colorbar` includes an `extend` kwarg, then that should be respected, regardless of whether the norm also has an `extend` kwarg.  It's only a subtle difference and small change from what you have.

shoyer commented:
Yes, I think I agree with this change. With xray, we did some work to make an external wrappers for pcolormesh, imshow and contourf that handles levels, extend and plot bounds that work identically for each plot type (although the underlying plot is produced differently). All this code exists in our [plotting module](https://github.com/xray/xray/blob/v0.6.1/xray/plot/plot.py) and you're free to adapt anything you like, of course (we use an Apache license).

fmaussion commented:
@efiring OK, I'll get back to this soon.

fmaussion commented:
I've incorporated @tacaswell requirements and added an image test ([see the image here](https://github.com/fmaussion/matplotlib/blob/fa5312a7566d6cc04da7d3251f14d781b0d93ea8/lib/matplotlib/tests/baseline_images/test_colors/boundarynorm_and_colorbar.png))

The failing test seems unrelated to my changes: it concerns only one single test environment? 

I'd like to update the examples for the documentation too but I will do this in a separate PR.

WeatherGod commented:
I restarted that test environment. Some of the tests can be a bit finicky on cloud environments.

fmaussion commented:
Hi again MPL gurus, @tacaswell wrote to mpl-devel that you have plans about generalizing the color map norms as a whole? Would that mean that this PR will soon be obsolete anyway?

I know that my PR is not of high importance and possibly also a bit awkward in the implementation, so I won't be offended if you choose to reject it ;)

tacaswell commented:
No, generalized norms are still some time out.

We can be very slow about review, it is nothing personal just a limit to how much bandwidth we have.

fmaussion commented:
Thanks Thomas, I didn't mean to be rude and I don't take it personally at all!

My remark was more related to the fact that earlier in the conversation (Sep 9th), @efiring mentioned that he had an idea for a more general solution to handle this. I tend to agree but I also have too little time to spend on it...

efiring commented:
@fmaussion The idea was half-baked, and I decided it probably wasn't going to provide a better solution than what you have here. That's why in my last message before this one I suggested you proceed with this PR.

fmaussion commented:
Folks, I guess there is no use in leaving this open for much longer. I don't want to use your time and besides me nobody seems to need this, so I guess we can close this?

efiring commented:
I think this just dropped off the radar at the point where I should have merged it; it looks good to me.  Apart from the need for a rebase to resolve the conflicts, do you have any reservations, @tacaswell?

tacaswell commented:
I still have slight reservations about the extra coupling, but not enough to hold the PR.

Can you do the rebase and add a `whats_new` entry?

fmaussion commented:
Thanks! Will do that tomorrow

fmaussion commented:
OK, this is ready for review!

jklymak commented:
@fmaussion any interest in seeing this through?  I think I understand what you are doing well enough to try to shepherd it through to final review.  It needs a rebase.  Sorry it moldered for so long!

fmaussion commented:
OK, rebased. Things have changed since 2015, so I have two questions:
- what happened to the "Colorbar only" example? Should I put it somewhere else, because now it is lost in a non-existing ``examples/api/`` folder
- same for ``what's new``: what's the recommended way to document a what's new?

efiring commented:
@fmaussion, I have a proposed simplification of the calculation.  May I push to your repo?  If so, I think there is something you need to click to add me as a collaborator.

fmaussion commented:
@efiring @jklymak I updated the PR with the following example which motivated this PR:

```python
import matplotlib.pyplot as plt
from matplotlib.colors import BoundaryNorm
import numpy as np

# Make the data
dx, dy = 0.05, 0.05
y, x = np.mgrid[slice(1, 5 + dy, dy),
                slice(1, 5 + dx, dx)]
z = np.sin(x) ** 10 + np.cos(10 + y * x) * np.cos(x)
z = z[:-1, :-1]

# Z roughly varies between -1 and +1
# my levels are chosen so that the color bar should be extended
levels = [-0.8, -0.5, -0.2, 0.2, 0.5, 0.8]
cmap = plt.get_cmap('PiYG')

# Before this change
plt.subplot(2, 1, 1)
norm = BoundaryNorm(levels, ncolors=cmap.N)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar(extend='both')
plt.axis([x.min(), x.max(), y.min(), y.max()])
plt.title('pcolormesh with extended colorbar')

# With the new keyword
norm = BoundaryNorm(levels, ncolors=cmap.N, extend='both')
plt.subplot(2, 1, 2)
im = plt.pcolormesh(x, y, z, cmap=cmap, norm=norm)
plt.colorbar()  # note that the colorbar is updated accordingly
plt.axis([x.min(), x.max(), y.min(), y.max()])
plt.title('pcolormesh with extended BoundaryNorm')

plt.show()
```

![bdnorm](https://user-images.githubusercontent.com/10050469/44574742-85fbd780-a78a-11e8-8735-19706f41a4a6.png)

I've also authorized edits from maintainers. Please feel free to make any change.

fmaussion commented:
Dear all, I am away from my laptop for the next three weeks - feel free to edit / merge as you wish.

jklymak commented on doc/users/whats_new/extend_kwarg_to_BoundaryNorm.rst:
````````diff
@@ -0,0 +1,46 @@
+cells.
+
+Example
+```````
````````
This example is pretty long for a rarely used kwarg - suggest just linking the real example...

jklymak replied:
Oh hmmm, there is no example.  Suggest this gets put in the appropriate section of `examples/`

fmaussion replied:
> This example is pretty long for a rarely used kwarg

I still wonder how people are doing that **without** this kwarg, but well. ;-)

>  Suggest this gets put in the appropriate section of examples/

will do

fmaussion commented:
While trying to add the concept to [this example](https://matplotlib.org/gallery/images_contours_and_fields/pcolormesh_levels.html#sphx-glr-gallery-images-contours-and-fields-pcolormesh-levels-py), I noticed that the coupling between ``norm`` and ``colorbar`` in the case of ``contourf`` does not work as expected :'-( : 

```python
import matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import BoundaryNorm
from matplotlib.ticker import MaxNLocator
import numpy as np


# make these smaller to increase the resolution
dx, dy = 0.05, 0.05

# generate 2 2d grids for the x & y bounds
y, x = np.mgrid[slice(1, 5 + dy, dy),
                slice(1, 5 + dx, dx)]

z = np.sin(x)**10 + np.cos(10 + y*x) * np.cos(x)

# x and y are bounds, so z should be the value *inside* those bounds.
# Therefore, remove the last value from the z array.
z = z[:-1, :-1]

# Z roughly varies between -1 and +1
# my levels are chosen so that the color bar should be extended
levels = [-0.8, -0.5, -0.2, 0.2, 0.5, 0.8]

# pick the desired colormap, sensible levels, and define a normalization
# instance which takes data values and translates those into levels.
cmap = plt.get_cmap('PiYG')
norm = BoundaryNorm(levels, ncolors=cmap.N, extend='both')

fig, (ax0, ax1) = plt.subplots(nrows=2)

im = ax0.pcolormesh(x, y, z, cmap=cmap, norm=norm)
fig.colorbar(im, ax=ax0)
ax0.set_title('pcolormesh with levels')


# contours are *point* based plots, so convert our bound into point
# centers
cf = ax1.contourf(x[:-1, :-1] + dx/2.,
                  y[:-1, :-1] + dy/2., z,
                  cmap=cmap, norm=norm)
fig.colorbar(cf, ax=ax1)
ax1.set_title('contourf with levels')

# adjust spacing between subplots so `ax1` title and `ax0` tick labels
# don't overlap
fig.tight_layout()

plt.show()
```

![d](https://user-images.githubusercontent.com/10050469/45832149-73c97680-bd01-11e8-8ab5-9dc44401f434.png)

Why is that so? Why should the signature of contourf and pcolormesh be different?

efiring commented:
Using BoundaryNorm with contourf makes no sense at all.  Maybe it should trigger a warning.  I suppose the alternative you are expecting is that contourf, when given a BoundaryNorm, would take its levels from that.  This seems to me like needless complexity, though.  

Contourf is fundamentally based on discrete levels; pcolormesh starts out with the opposite point of view, that it is representing a continuum.  BoundaryNorm provides a discretization mechanism for pcolormesh, but the discretization is built in to contourf.

fmaussion commented:
Thanks for the quick reply. I will write a new dedicated example then

efiring commented:
@jklymak and @fmaussion As far as you can see, is anything needed beyond a rebase?  Are there outstanding questions still to be resolved? It would be nice to get this finished and merged ASAP so it doesn't sit around for a few more years.

jklymak commented:
Away from a real computer for a few days so don’t wait for me!

efiring commented:
@fmaussion I would like to get this merged.  Are you available to resolve the conflict and address anything else that is pending?

efiring commented:
The present conflict is that this PR is deleting tutorials/colors/colorbar_only.py.  We definitely don't want to do that.
Apart from that, what's missing is the entry in "examples" or "tutorials" illustrating the new feature here.

jklymak commented:
@fmaussion This looks to be done, but needs a rebase and the colorbar_only tutorial re-instated!  Did you want to tackle it, or should one of the devs?

efiring commented:
It also looks like the example from api_changes needs to be moved to the reinstated colorbar_only.py tutorial.

efiring commented:
Closing, since I have rebased and moved this to #17534.
